tests/test_networks.py

```diff
@@ -0,0 +1,356 @@
+        (dict(scheme='http', host='example.net', port='1234'), 'http://example.net:1234'),
+    ],
+)
+def test_build_url(kwargs, expected):
+    assert AnyUrl(None, **kwargs) == expected
+
+
+def test_son():
+    class Model(BaseModel):
+        v: HttpUrl
+
+    m = Model(v='http://foo@example.net')
+    assert m.json() == '{"v": "http://foo@example.net"}'
+    assert m.schema() == {
+        'title': 'Model',
+        'type': 'object',
+        'properties': {'v': {'title': 'V', 'minLength': 1, 'maxLength': 2083, 'type': 'string', 'format': 'uri'}},
+        'required': ['v'],
+    }
+
+
+@pytest.mark.skipif(not email_validator, reason='email_validator not installed')
+@pytest.mark.parametrize(
+    'value,name,email',
+    [
+        ('foobar@example.com', 'foobar', 'foobar@example.com'),
+        ('s@muelcolvin.com', 's', 's@muelcolvin.com'),
+        ('Samuel Colvin <s@muelcolvin.com>', 'Samuel Colvin', 's@muelcolvin.com'),
+        ('foobar <foobar@example.com>', 'foobar', 'foobar@example.com'),
+        (' foo.bar@example.com', 'foo.bar', 'foo.bar@example.com'),
+        ('foo.bar@example.com ', 'foo.bar', 'foo.bar@example.com'),
+        ('foo BAR <foobar@example.com >', 'foo BAR', 'foobar@example.com'),
+        ('FOO bar   <foobar@example.com> ', 'FOO bar', 'foobar@example.com'),
+        ('<FOOBAR@example.com> ', 'FOOBAR', 'foobar@example.com'),
+        ('ñoñó@example.com', 'ñoñó', 'ñoñó@example.com'),
+        ('我買@example.com', '我買', '我買@example.com'),
+        ('甲斐黒川日本@example.com', '甲斐黒川日本', '甲斐黒川日本@example.com'),
+        (
+            'чебурашкаящик-с-апельсинами.рф@example.com',
+            'чебурашкаящик-с-апельсинами.рф',
+            'чебурашкаящик-с-апельсинами.рф@example.com',
+        ),
+        ('उदाहरण.परीक्ष@domain.with.idn.tld', 'उदाहरण.परीक्ष', 'उदाहरण.परीक्ष@domain.with.idn.tld'),
+        ('foo.bar@example.com', 'foo.bar', 'foo.bar@example.com'),
+        ('foo.bar@exam-ple.com ', 'foo.bar', 'foo.bar@exam-ple.com'),
+        ('ιωάννης@εεττ.gr', 'ιωάννης', 'ιωάννης@εεττ.gr'),
+    ],
+)
+def test_address_valid(value, name, email):
+    assert validate_email(value) == (name, email)
+
+
+@pytest.mark.skipif(not email_validator, reason='email_validator not installed')
+@pytest.mark.parametrize(
+    'value',
+    [
+        'f oo.bar@example.com ',
+        'foo.bar@exam\nple.com ',
+        'foobar',
+        'foobar <foobar@example.com',
+        '@example.com',
+        'foobar@.example.com',
+        'foobar@.com',
+        'foo bar@example.com',
+        'foo@bar@example.com',
+        '\n@example.com',
+        '\r@example.com',
+        '\f@example.com',
+        ' @example.com',
+        '\u0020@example.com',
+        '\u001f@example.com',
+        '"@example.com',
+        '\"@example.com',
+        ',@example.com',
+        'foobar <foobar<@example.com>',
+    ],
+)
+def test_address_invalid(value):
+    with pytest.raises(ValueError):
+        validate_email(value)
```

Comment:
What are the error messages like on these? I wonder how many of them actually end up being valid as far as the RFCs in play are concerned? *sigh*

Reply:
This hasn't changed, just moved the tests. The error message will be given the [email-validator](https://pypi.org/project/email-validator/) package.